dev/tools/controllerbuilder/pkg/codegen/typegenerator.go

```diff
@@ -0,0 +1,284 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codegen
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	protoapi "github.com/GoogleCloudPlatform/k8s-config-connector/dev/tools/controllerbuilder/pkg/protoapi"
+
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"k8s.io/klog/v2"
+)
+
+type TypeGenerator struct {
+	generatorBase
+	filter FilterFunc
+}
+
+func NewTypeGenerator(filter FilterFunc) *TypeGenerator {
+	g := &TypeGenerator{
+		filter: filter,
+	}
+	g.generatorBase.init()
+	return g
+}
+
+type FilterFunc func(msg protoreflect.MessageDescriptor) bool
+
+func (v *TypeGenerator) VisitProto(api *protoapi.Proto) error {
+	sortedFiles := api.SortedFiles()
+	for _, f := range sortedFiles {
+		v.visitFile(f)
+	}
+	return nil
+}
+
+func (g *TypeGenerator) visitFile(f protoreflect.FileDescriptor) {
+	for _, msg := range sorted(f.Messages()) {
+		g.visitMessage(msg)
+	}
+
+	krmVersion := "v1beta1"
```

Comment:
Shall we set the default to `v1alpha1`? Assuming all direct resources shall start from Alpha.

Reply:
Good point!

So we have the actual version in a flag passed to the command.  I changed the logic around the filter function so that it now returns the go path in which to write messages, meaning we use the version the user asked for :-)